lib/linked_list.rb

```diff
@@ -7,39 +7,72 @@ def initialize
     @head = nil
   end
 
-  # Time complexity - ?
-  # Space complexity - ?
+  # Time complexity - O(1)
+  # Space complexity - O(1)
   def add_first(data)
-
+    @head = Node.new(data, head)
   end
 
-  # Time complexity - ?
-  # Space complexity - ?
+  # Time complexity - O(1)
+  # Space complexity - O(1)
   def get_first
-
+    return nil if @head.nil?
+    return @head.data
   end
 
-  # Time complexity - ?
-  # Space complexity - ?
+  # Time complexity - O(n)
+  # Space complexity - O(1)
   def length
-    return 0
+    current = @head
+    counter = 0
+    until current.nil?
+      counter += 1
+      current = current.next
+    end
+    return counter
   end
 
-  # Time complexity - ?
-  # Space complexity - ?
+  # Time complexity - O(n)
+  # Space complexity - O(1)
   def add_last(data)
-
+    new_node = Node.new(data)
+    
+    if @head == nil 
+      @head = new_node
+    else
+      current = @head
+      until current.next.nil?
+        current = current.next
+      end
+      current.next = new_node
+    end
+    
   end
 
-  # Time complexity - ?
-  # Space complexity - ?
+  # Time complexity - O(n)
+  # Space complexity - O(1)
   def get_last
-
+    current = @head
```

Comment:
What if `head` is `nil`?

```suggestion
    return nil if head.nil?
    current = head
```